app/services/openid_connect_redirector.rb

```diff
@@ -73,11 +73,25 @@ def validate_redirect_uri_matches_sp_redirect_uri
     errors.add(error_attr, t('openid_connect.authorization.errors.redirect_uri_no_match'))
   end
 
+  # rubocop:disable Metrics/AbcSize
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/MethodLength
   def redirect_uri_matches_sp_redirect_uri?
-    redirect_uri.present? &&
-      service_provider.active? &&
-      service_provider.redirect_uris.any? do |sp_redirect_uri|
-        redirect_uri.start_with?(sp_redirect_uri)
-      end
+    return unless redirect_uri.present? && service_provider.active?
+    parsed_redirect_uri = URI(redirect_uri)
+    service_provider.redirect_uris.any? do |sp_redirect_uri|
+      parsed_sp_redirect_uri = URI(sp_redirect_uri)
+
+      parsed_redirect_uri.scheme == parsed_sp_redirect_uri.scheme &&
+        parsed_redirect_uri.port == parsed_sp_redirect_uri.port &&
+        parsed_redirect_uri.host == parsed_sp_redirect_uri.host &&
+        parsed_redirect_uri.path.start_with?(parsed_sp_redirect_uri.path)
```

Comment:
Is the expectation that an SP can make a request with a redirect_uri that doesn't exactly match one that is registered? Would it make sense to enforce exact matches?

Reply:
I had the same thought. Is the point here to provide some flexibility for integrating partners for them to change their internal implementation details without having to get us to redeploy?

Reply:
From my analysis of service_provider_requests it appears every agency is sending exact matches.  I don't see any reason why we can't start enforcing it.  @brodygov analyzed the production logs.  I'm not sure what his results were with respect to exact matching.

Reply:
Updating Service Provider attributes can be done independently of a deploy, so even more reason to enforce exact matches. I have a refactor branch in the works that removes some redundant validations. I can update it to enforce exact matches.

Reply:
My recollection is that we did prefix matching because at the time we only allowed a single redirect URI, so that was an easy way to allow support for multiple paths. Later we added support for multiple redirect URIs, so I believe the prefix matching is no longer needed.